bookkeeper-server/src/main/java/org/apache/bookkeeper/meta/MetadataDrivers.java

```diff
@@ -0,0 +1,290 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Sets;
+import java.net.URI;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.ToString;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.util.ReflectionUtils;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * A driver manager for managing a set of metadata drivers.
+ *
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+@Slf4j
+public final class MetadataDrivers {
+
+    static final String ZK_CLIENT_DRIVER_CLASS = "org.apache.bookkeeper.meta.zk.ZKMetadataClientDriver";
+    static final String ZK_BOOKIE_DRIVER_CLASS = "org.apache.bookkeeper.meta.zk.ZKMetadataBookieDriver";
+    static final String BK_METADATA_CLIENT_DRIVERS_PROPERTY = "bookkeeper.metadata.client.drivers";
+    static final String BK_METADATA_BOOKIE_DRIVERS_PROPERTY = "bookkeeper.metadata.bookie.drivers";
+
+    @ToString
+    static class MetadataClientDriverInfo {
+
+        final Class<? extends MetadataClientDriver> driverClass;
+        final String driverClassName;
+
+        MetadataClientDriverInfo(Class<? extends MetadataClientDriver> driverClass) {
+            this.driverClass = driverClass;
+            this.driverClassName = driverClass.getName();
+        }
+
+    }
+
+    @ToString
+    static class MetadataBookieDriverInfo {
+
+        final Class<? extends MetadataBookieDriver> driverClass;
+        final String driverClassName;
+
+        MetadataBookieDriverInfo(Class<? extends MetadataBookieDriver> driverClass) {
+            this.driverClass = driverClass;
+            this.driverClassName = driverClass.getName();
+        }
+
+    }
+
+    @Getter(AccessLevel.PACKAGE)
+    private static final ConcurrentMap<String, MetadataClientDriverInfo> clientDrivers;
+    @Getter(AccessLevel.PACKAGE)
+    private static final ConcurrentMap<String, MetadataBookieDriverInfo> bookieDrivers;
+    private static boolean initialized = false;
+
+    static {
+        clientDrivers = new ConcurrentHashMap<>();
+        bookieDrivers = new ConcurrentHashMap<>();
+        initialize();
+    }
+
+    static void initialize() {
+        if (initialized) {
+            return;
+        }
+        loadInitialDrivers();
+        initialized = true;
+        log.info("BookKeeper metadata driver manager initialized");
+    }
+
+    @VisibleForTesting
+    static void loadInitialDrivers() {
+        loadInitialClientDrivers();
+        loadInitialBookieDrivers();
+    }
+
+    private static void loadInitialClientDrivers() {
+        Set<String> driverList = Sets.newHashSet();
+
+        // add default zookeeper based driver
+        driverList.add(ZK_CLIENT_DRIVER_CLASS);
+
+        // load drivers from system property
+        String driversStr = System.getProperty(BK_METADATA_CLIENT_DRIVERS_PROPERTY);
```

Comment:
I think it should be better to use standard Java Service Provider API. It is 'standard' and JPMS aware

Reply:
This pattern was following JDBC in java 7. I can change to use ServiceLoader as we are in java 8 (the way that JDBC is using in java 8).

Reply:
actually I think jdbc can still load drivers from [system properties](http://hg.openjdk.java.net/jdk8/jdk8/jdk/file/tip/src/share/classes/java/sql/DriverManager.java#l572). 

So I would keep this setting here and add the ServiceLoader support later once we have a clean module definition, unless there are other major concerns on current approach.

Reply:
Okay for the system property for now